server/service/src/sync/translations/category.rs

```diff
@@ -0,0 +1,135 @@
+use crate::sync::sync_serde::empty_str_as_option_string;
+use repository::{
+    category_row::{CategoryRow, CategoryRowDelete},
+    StorageConnection, SyncBufferRow,
+};
+use serde::{Deserialize, Serialize};
+
+use super::{PullTranslateResult, SyncTranslation};
+
+#[allow(non_camel_case_types)]
+#[derive(Deserialize, Serialize)]
+pub enum LegacyItemType {
+    non_stock,
+    service,
+    general,
+}
+
+#[allow(non_snake_case)]
+#[derive(Deserialize, Serialize)]
+pub struct LegacyItemCategoryRow {
+    ID: String,
+    Description: String,
+    sort_order: i32,
+    #[serde(deserialize_with = "empty_str_as_option_string")]
+    parent_ID: Option<String>,
+}
+
+#[allow(non_snake_case)]
+#[derive(Deserialize, Serialize)]
+pub struct LegacyItemCategoryLevel1Row {
+    ID: String,
+    Description: String,
+    sort_order: i32,
+}
+
+// Needs to be added to all_translators()
+#[deny(dead_code)]
+pub(crate) fn boxed() -> Box<dyn SyncTranslation> {
+    Box::new(CategoryTranslation)
+}
+
+pub(super) struct CategoryTranslation;
+impl SyncTranslation for CategoryTranslation {
+    fn table_name(&self) -> &str {
+        "item_category"
+    }
+
+    fn pull_dependencies(&self) -> Vec<&str> {
+        vec![]
+    }
+
+    fn should_translate_from_sync_record(&self, row: &SyncBufferRow) -> bool {
+        match row.table_name.as_str() {
+            "item_category" | "item_category_level2" | "item_category_level1" => true,
+            _ => false,
+        }
+    }
+
+    fn try_translate_from_upsert_sync_record(
+        &self,
+        _: &StorageConnection,
+        sync_record: &SyncBufferRow,
+    ) -> Result<PullTranslateResult, anyhow::Error> {
+        // Try to parse as a child category (item_category or item_category_level2)
```

Comment:
I feel like this probably, can match both?

If you put `#[serde(default)]` on parent_id, you should be able to just use one legacy type?
Am I right the only difference is the existence of the parent_id?

Reply:
Oo nice spot, thank you!